velox/functions/sparksql/DecimalArithmetic.cpp

```diff
@@ -14,498 +14,360 @@
-  // the scaled value will not exceed the maximum of long decimal.
-  return delta <= 0 ? in : in * velox::DecimalUtil::kPowersOfTen[delta];
+// Return precision of `aType` and `bType`, set the `aScale` and `bScale`.
+std::pair<int, int> getDecimalPrecisionScales(
+    const Type& aType,
+    const Type& bType,
+    uint8_t& aScale,
+    uint8_t& bScale) {
+  auto [aPrecision, aScaleTmp] = getDecimalPrecisionScale(aType);
+  auto [bPrecision, bScaleTmp] = getDecimalPrecisionScale(bType);
+  aScale = aScaleTmp;
+  bScale = bScaleTmp;
+  return {aPrecision, bPrecision};
 }
 
-// Scales up the whole part to result scale, and combine it with fraction part
-// to produce a full result for decimal add. Checks whether the result
-// overflows.
-template <typename T>
-inline T
-decimalAddResult(T whole, T fraction, uint8_t resultScale, bool& overflow) {
-  T scaledWhole = DecimalUtil::multiply<T>(
-      whole, velox::DecimalUtil::kPowersOfTen[resultScale], overflow);
-  if (FOLLY_UNLIKELY(overflow)) {
-    return 0;
-  }
-  const auto result = scaledWhole + fraction;
-  if constexpr (std::is_same_v<T, int64_t>) {
-    overflow = (result > velox::DecimalUtil::kShortDecimalMax) ||
-        (result < velox::DecimalUtil::kShortDecimalMin);
-  } else {
-    overflow = (result > velox::DecimalUtil::kLongDecimalMax) ||
-        (result < velox::DecimalUtil::kLongDecimalMin);
+struct DecimalAddSubtractBase {
+ public:
+  void initializeBase(const std::vector<TypePtr>& inputTypes) {
+    auto aType = inputTypes[0];
+    auto bType = inputTypes[1];
+    auto [aPrecision, bPrecision] = getDecimalPrecisionScales(
+        *(inputTypes[0]), *(inputTypes[1]), aScale_, bScale_);
+    aRescale_ = computeRescaleFactor(aScale_, bScale_);
+    bRescale_ = computeRescaleFactor(bScale_, aScale_);
+    auto [rPrecision, rScale] =
+        computeResultPrecisionScale(aPrecision, aScale_, bPrecision, bScale_);
+    rPrecision_ = rPrecision;
+    rScale_ = rScale;
   }
-  return result;
-}
 
-// Reduces the scale of input value by 'delta'. Returns the input value if delta
-// is not positive.
-template <typename T>
-inline static T reduceScale(T in, int32_t delta) {
-  if (delta <= 0) {
-    return in;
-  }
-  T result;
-  bool overflow;
-  const auto scaleFactor = velox::DecimalUtil::kPowersOfTen[delta];
-  if constexpr (std::is_same_v<T, int64_t>) {
-    VELOX_DCHECK_LE(
-        scaleFactor,
-        std::numeric_limits<int64_t>::max(),
-        "Scale factor should not exceed the maximum of int64_t.");
+  // Adds the values 'a' and 'b' and stores the result in 'r'. To align the
+  // scales of inputs, the value with the smaller scale is rescaled to the
+  // larger scale. 'aRescale' and 'bRescale' are the rescale factors needed to
+  // rescale 'a' and 'b'. 'rPrecision' and 'rScale' are the precision and scale
+  // of the result.
+  template <typename TResult, typename A, typename B>
+  inline bool applyAdd(TResult& r, A a, B b) {
```

Comment:
Perhaps we don't need the inline keyword as member functions (including static ones) implemented inside a class or struct are automatically inline.